### PR TITLE
PebbleDB Iterator Fix

### PR DESCRIPTION
**Pull request by Kbhat1**
## Describe your changes and provide context
- Separates out logic for forward and reverse iteration
- During iterator initialization, makes sure to call `Next` if not on a key which is valid (version <= requested version)
- For Forward and Reverse case, adds a check at end to verify if the key has a version <= requested version, if not, will recursively call `Next` again. Previous logic did not guarantee in such edge cases that a Next call would move the iterator either to a key which is valid or exhaust it

## Testing performed to validate your change
- Unit tests added
- Testing on node + integration tests

### CODE REVIEW

File: ss/pebbledb/iterator.go [hunk 1]
```diff
@@ -115,23 +115,22 @@ func (itr *iterator) Value() []byte {
 	return slices.Clone(val)
 }
 
-func (itr *iterator) Next() {
+func (itr *iterator) NextForward() {
```
Comment (yzang2019) [inline ss/pebbledb/iterator.go:118]: We should make this function private to avoid caller call it directly right?

File: ss/pebbledb/iterator.go [hunk 2]
```diff
@@ -115,23 +115,22 @@ func (itr *iterator) Value() []byte {
...
-	var next bool
-	if itr.reverse {
-		// Since PebbleDB has no PrevPrefix API, we must manually seek to the next
-		// key that is lexicographically less than the current key.
-		next = itr.source.SeekLT(MVCCEncode(currKey, 0))
-	} else {
-		// move the cursor to the next key
-		next = itr.source.NextPrefix()
-	}
+	fmt.Printf("currKey %s\n", string(currKey))
```
Comment (alexanderbez) [inline ss/pebbledb/iterator.go:131]: There's still debug lines in this PR

Comment (Kbhat1) [reply ss/pebbledb/iterator.go:131]: Yea, as mentioned I'm leaving these in for now. I'll change this to draft and re-open when more tests / sanity checks are complete and this is all cleaned up.

File: ss/pebbledb/iterator.go [hunk 3]
```diff
@@ -73,10 +60,23 @@ func newPebbleDBIterator(src *pebble.Iterator, prefix, mvccStart, mvccEnd []byte
...
+			// MVCC value.
+			panic(fmt.Sprintf("invalid PebbleDB MVCC value: %s", itr.source.Key()))
+		}
+
+		curKeyVersionDecoded, err := decodeUint64Ascending(currKeyVersion)
+		if err != nil {
+			panic(err)
+		}
+
+		// Edge case: the first key has multiple versions so we need to do a seek
```
Comment (alexanderbez) [inline ss/pebbledb/iterator.go:76]:
```suggestion
		// if the first key has multiple versions, we need to do a seek
```

**Comment by alexanderbez** [comment]
Will review this ASAP

File: ss/pebbledb/iterator.go [hunk 4]
```diff
@@ -154,7 +153,7 @@ func (itr *iterator) Next() {
 		// append the current iterator key to the prefix and seek to that key.
 		itr.valid = itr.source.SeekLT(MVCCEncode(nextKey, itr.version+1))
 
-		tmpKey, _, ok := SplitMVCCKey(itr.source.Key())
+		tmpKey, tmpKeyVersion, ok := SplitMVCCKey(itr.source.Key())
```
Comment (alexanderbez) [inline ss/pebbledb/iterator.go:156]: nit: this is critical code. I would add a godoc above stating what/why we're doing this.

File: ss/pebbledb/iterator.go [hunk 5]
```diff
@@ -166,17 +165,102 @@ func (itr *iterator) Next() {
...
+
 			} else {
 				itr.valid = false
 				return
 			}
 		}
 
+		tmpKeyVersionDecoded, err := decodeUint64Ascending(tmpKeyVersion)
+		if err != nil {
+			panic(err)
```
Comment (alexanderbez) [inline ss/pebbledb/iterator.go:186]: IMO we should not panic here, instead return like we do in other cases.

File: ss/pebbledb/iterator.go [hunk 6]
```diff
@@ -73,10 +60,25 @@ func newPebbleDBIterator(src *pebble.Iterator, prefix, mvccStart, mvccEnd []byte
...
+		curKeyVersionDecoded, err := decodeUint64Ascending(currKeyVersion)
+		if err != nil {
+			panic(err)
+		}
+
+		if curKeyVersionDecoded > itr.version {
+			itr.Next()
+		} else {
+			// Seek to the largest version of that key <= requested iterator version
+			itr.valid = itr.source.SeekLT(MVCCEncode(currKey, itr.version+1))
```
[the remark this answers is not in the document]
Comment (yzang2019) [reply ss/pebbledb/iterator.go:80]: In this case it guarantees that curKeyVersionDecoded <= target version, which means there must exist one version that SeekLT will return for the current key, which I think should be safe

File: ss/pebbledb/iterator.go (hunk 4 above)
Comment (yzang2019) [reply ss/pebbledb/iterator.go:156]: +1

File: ss/pebbledb/iterator.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (alexanderbez) [reply ss/pebbledb/iterator.go:80]: We should update the godoc comment with what you stated IMO 👍

**Review by yzang2019 — APPROVED** [review]
